File: framework/languages/dotnet/e2e/SpotService/Client/Scenarios/SmD10Scenario.cs
using SpotService.Client;
using SpotService.Shared;
using Systems.Zlink.Stream.Connector.Contracts;

namespace SpotService.Client.Scenarios;

internal static class SmD10Scenario
{
    public static async Task RunAsync(string sessionAStreamEndpoint, string sessionBStreamEndpoint)
    {
        await using var congested = await ConnectAndAuthWithRetryAsync(
            sessionAStreamEndpoint,
            new AuthReq("actor-sm-d10-congested", "stream backpressure", "play-a"),
            maxReceivedMessages: 1);
        await using var isolated = await ConnectAndAuthWithRetryAsync(
            sessionBStreamEndpoint,
            new AuthReq("actor-sm-d10-isolated", "stream backpressure peer", "play-b"));

        for (var index = 0; index < 8; index++)
        {
            var reply = await congested.Request(new ActorPushReq($"burst-{index}"))
                .PacketName("ActorPushReq")
                .Async<ActorPingReply>();
            ScenarioAssert.That(reply.ActorId == "actor-sm-d10-congested", "SM-D10 congested reply actor mismatch.");
        }

        ScenarioAssert.That(
            congested.ReceivedCount("ActorPushNotify") <= 1,
            "SM-D10 expected bounded received-message queue for congested session.");
        var retained = await congested.WaitFor<ActorPushNotify>()
            .Where(message => message.Payload.ActorId == "actor-sm-d10-congested")
            .Timeout(TimeSpan.FromSeconds(2))
            .Async();
        ScenarioAssert.That(retained.Payload.Value == "burst-7", "SM-D10 expected the newest congested push to be retained.");

        var stillAlive = await congested.Request(new ActorPingReq("after-backpressure"))
            .PacketName("ActorPingReq")
            .Async<ActorPingReply>();
        ScenarioAssert.That(stillAlive.ActorId == "actor-sm-d10-congested", "SM-D10 congested session stopped routing.");
        ScenarioAssert.That(stillAlive.Value == "after-backpressure", "SM-D10 congested session reply mismatch.");

        var isolatedPush = isolated.WaitFor<ActorPushNotify>().Async().AsTask();
        var isolatedReply = await isolated.Request(new ActorPushReq("isolated-push"))
            .PacketName("ActorPushReq")
            .Async<ActorPingReply>();
        var isolatedNotify = await isolatedPush;
        ScenarioAssert.That(isolatedReply.ActorId == "actor-sm-d10-isolated", "SM-D10 isolated reply actor mismatch.");
        ScenarioAssert.That(isolatedNotify.Payload.ActorId == "actor-sm-d10-isolated", "SM-D10 isolated push actor mismatch.");
        ScenarioAssert.That(isolatedNotify.Payload.Value == "isolated-push", "SM-D10 isolated session push mismatch.");

        Console.WriteLine("operation SpotService.sm-d10 passed");
    }

    static async Task<IZlinkStreamConnector> ConnectAndAuthWithRetryAsync(
        string endpoint,
        AuthReq auth,
        int maxReceivedMessages = 1024)
    {
        var deadline = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(10);
        Exception? last = null;
        while (DateTimeOffset.UtcNow < deadline)
        {
            var client = CreateClient(endpoint, maxReceivedMessages);
            try
            {
                await client.Connect.Async();
                await client.Request(auth)
                    .PacketName("AuthReq")
                    .Async<AuthReply>();
                return client;
            }
            catch (Exception ex) when (ex is ZlinkStreamException or TimeoutException)
            {
                last = ex;
                await client.DisposeAsync();
                await Task.Delay(200);
            }
        }

        throw new InvalidOperationException(
            last is null ? $"Actor auth did not become routable: {auth.ActorId}" : $"Actor auth did not become routable: {auth.ActorId}. Last error: {last.Message}",
            last);
    }

    static IZlinkStreamConnector CreateClient(string endpoint, int maxReceivedMessages)
    {
        ScenarioAssert.That(!string.IsNullOrWhiteSpace(endpoint), "stream endpoint is required.");
        return ZlinkStreamConnectorFactory.Create(new ZlinkStreamConnectorOptions
        {
            Endpoint = new Uri(endpoint),
            ConnectTimeout = TimeSpan.FromSeconds(5),
            RequestTimeout = TimeSpan.FromSeconds(5),
            Heartbeat = new ZlinkStreamHeartbeatOptions { Enabled = false },
            DispatchMode = ZlinkStreamDispatchMode.Immediate,
            MaxReceivedMessages = maxReceivedMessages,
        });
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Client/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/client/scenarios/SmD10Scenario.kt
package systems.zlink.e2e.kotlin.spotservice.client.scenarios

import systems.zlink.framework.kotlin.*

import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.Env
import systems.zlink.e2e.kotlin.spotservice.client.support.createStreamConnector
import systems.zlink.e2e.kotlin.spotservice.client.support.ensure
import systems.zlink.stream.connector.ZLinkStreamDispatchMode

internal object SmD10Scenario {
    suspend fun run() {
        val congestedActorId = "actor-sm-d10-congested"
        val isolatedActorId = "actor-sm-d10-isolated"
        val congested = createStreamConnector(
            Env.get("ZLINK_KOTLIN_E2E_STREAM_A_ENDPOINT"),
            ZLinkStreamDispatchMode.MANUAL,
            1,
        )
        val isolated = createStreamConnector(Env.get("ZLINK_KOTLIN_E2E_STREAM_B_ENDPOINT"))
        try {
            val congestedProfile = Contracts.ActorProfile("Backpressure", 10, listOf("congested"))
            val isolatedProfile = Contracts.ActorProfile("Backpressure Peer", 10, listOf("isolated"))
            congested.connect().await()
            congested
                .request(Contracts.ActorAuthReq(congestedActorId, congestedProfile))
                .await<Contracts.ActorAuthRes>()
            isolated.connect().await()
            isolated
                .request(Contracts.ActorAuthReq(isolatedActorId, isolatedProfile))
                .await<Contracts.ActorAuthRes>()

            val retained = coroutineScope {
                val retainedPush = async(start = CoroutineStart.UNDISPATCHED) {
                    congested.waitFor<Contracts.ActorPushNotify>().await()
                }
                for (index in 0 until 8) {
                    val reply = congested
                        .request(Contracts.ActorEchoReq("burst-$index", 10, congestedProfile))
                        .await<Contracts.ActorEchoRes>()
                    ensure(reply.actorId == congestedActorId, "SM-D10 congested reply actor mismatch")
                }
                ensure(congested.receivedCount("ActorPushNotify") <= 1, "SM-D10 congested queue retained too many pushes")
                congested.dispatch().await()
                retainedPush.await().payload()
            }
            ensure(retained.actorId == congestedActorId, "SM-D10 retained push actor mismatch")
            ensure(retained.value == "push:burst-7", "SM-D10 expected newest congested push to be retained")

            val stillAlive = congested
                .request(Contracts.ActorEchoReq("after-backpressure", 10, congestedProfile))
                .await<Contracts.ActorEchoRes>()
            ensure(stillAlive.actorId == congestedActorId, "SM-D10 congested session stopped routing")
            ensure(stillAlive.value == "entry:after-backpressure", "SM-D10 congested session reply mismatch")
            congested.dispatch().await()

            val (isolatedReply, isolatedNotify) = coroutineScope {
                val isolatedPush = async(start = CoroutineStart.UNDISPATCHED) {
                    isolated.waitFor<Contracts.ActorPushNotify>().await()
                }
                val isolatedReply = isolated
                    .request(Contracts.ActorEchoReq("isolated-push", 10, isolatedProfile))
                    .await<Contracts.ActorEchoRes>()
                isolatedReply to isolatedPush.await().payload()
            }
            ensure(isolatedReply.actorId == isolatedActorId, "SM-D10 isolated reply actor mismatch")
            ensure(isolatedNotify.actorId == isolatedActorId, "SM-D10 isolated push actor mismatch")
            ensure(isolatedNotify.value == "push:isolated-push", "SM-D10 isolated session push mismatch")

            println("scenario SM-D10 passed")
        } finally {
            try {
                congested.close().await()
            } catch (_: Exception) {
            }
            try {
                isolated.close().await()
            } catch (_: Exception) {
            }
        }
    }
}
